Hackable Station Networks [WIP]

nullbear opened the pull request:
**As supermichael777 said, "what is this?":**

I'm currently working on a station-wide intranet system to be used between machines, allowing for a lot more interactivity by players and the ai and between different machines. It's intended to work alongside many machines. It allows for players to implement various automated mechanics, such as virology security airlocks, and to tinker with machines a lot more intimately, typically having access to more functions via the network than you would via cutting and pulsing wires. An example of a device that could be implemented by a player, is a button which bolts all doors connected to the same network in the event of an emergency. Or a protocol which automatically lifts access restrictions for security and medical in the case of a DELTA or RED alert.

At the moment, i've mostly handled commands for accessing, traversing, and modifying networks, and have yet to polish up an interface for it, or add any interactivity between machinery and networks.

For a more simplistic way of looking at it. A machines network is like a virtual version of the machines wires. Instead of being accessed with a multitool or wirecutters, you use the hacktool. The primary difference being that you're given TONS more control than just 'cut/pulse', and that many networks can be accessed remotely, or through connected networks.

It's intended to be used primarily for automation, machine linking, traitors, and silicons.

---

**Upgrades:**

Standard, non-upgraded hacktools, are typically quite bad at well, 'hacking', and are designed more for navigating and operating on networks, than they are for probing and hacking into secure networks. But have no fear, as there's a fair number of different software upgrades that can be installed to improve its functionality.

STEALTH software, reduces the chances of being detected or locked out by the security systems.

PROBE software, allows you to access hidden networks and commands, but increases the chances of being detected by security.

BRUTE software, allows you to bypass network encryption keys and passwords, but will typically cause you to be detected by the security systems unless stealth software is also installed.

BOOST software, boosts the signal of the hacktool, allowing it to connect to networks from further away. Remote station networks become accessible from off-station. Remote area networks become accessible from anywhere on the station. And non-remote networks become accessible from within the same area.

BYPASS software, allows you to enable bypass mode, which bypasses security and network lockouts for a short period of time, but will alert the security systems loudly.

SPEED | HYPER software. Increasing hacktool software speeds by 1.5x, or 2x respectively. They are best used simultaneously, for a 3x speed boost.

SECURE software, increases the time required for security systems launching a counterattack. Very good if you don't want to get caught by a tattletale A.I.

---

**Network security:**

Network security systems are built to be modular, with many hooks built into every other network command. Some security systems may be more forgiving than others, where they will allow gentle network probing and failed password attempts. Or they may be rather harsh, locking out the system for several minutes, bolting down the room, sounding alarms and notifying the AI, when so much as a 'help' command is entered.

Security can be battled using software upgrades such as STEALTH or BYPASS, allowing you to avoid being detected in the first place, or alternatively, to bypass the security system for a short time, at the cost of triggering louder countermeasures. 

Some Security systems may also attempt to launch a counter-attack, pulling information from the hacktool, and potentially preventing it from being used temporarily. This takes a bit of time, and you will have a few seconds to react. The information pulled by the security system will usually contain the hacktools location, and if an ID is inserted into it, the designation of that ID. If you suspect that a security system is launching a counterattack, the best course of action is to remove your ID from the hacktool, and retreat to a crowded area, or distance yourself from the hacktool. (if the detective finds it, it could still have your prints. And the ai might find it suspicious if they see you returning to pick it up.)

When a security alert goes off, the ai will receive a ping to the location. For this reason it's typically safer to hack from as far as possible from the network source, away from any cameras, and via a proxy.

---

RemieRichards commented on code/__HELPERS/text.dm:
```diff
@@ -425,3 +425,18 @@ var/list/binary = list("0","1")
 	for(var/x in 1 to length(t))
 		. += copytext(t,x,x+1)
+
+proc/explode_text(text, separator = " ")
```
exists as ```splittext()```

nullbear replied:
thanks, i was looking for it all over the damned place.

RemieRichards commented on code/__HELPERS/text.dm:
```diff
@@ -425,3 +425,18 @@ var/list/binary = list("0","1")
+	var/i
+	var/oldi = 1
+	var/list/L = new/list()
+	for(i=1, i<=length(text)+1, i++)
```
```for(var/i in 1 to length(text)+1)```

RemieRichards commented on code/__HELPERS/text.dm:
```diff
@@ -425,3 +425,18 @@ var/list/binary = list("0","1")
+	var/list/L = new/list()
+	for(i=1, i<=length(text)+1, i++)
+		if(copytext(text, i, i+length(separator)) == separator) // We found a separator
+			L.Add(copytext(text, oldi, i))
```
```+=```, not ```Add()```

RemieRichards commented on code/__HELPERS/text.dm:
```diff
@@ -425,3 +425,18 @@ var/list/binary = list("0","1")
+		if(copytext(text, i, i+length(separator)) == separator) // We found a separator
+			L.Add(copytext(text, oldi, i))
+			oldi = i+length(separator)
+	L.Add(copytext(text, oldi, length(text)+1)) // Adds the last segment
```
```+=```, not ```Add()```

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+	..()
+	if(A.args.len == 0)
+		feedback = "<span class='notice'>Available Network Commands: </span>"
+		for(var/datum/network_command/NC in N.commands)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+			if(!NC.stealth || ishacktool(H))
+				feedback += "\n <span class='notice'>[NC.trigger]</span>"
+		return feed(H)
+	for(var/datum/network_command/NC in N.commands)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+	if(N.sub.len)
+		var/list/datum/network/SN = list()
+		feedback = "<span class='notice'>Available Subnetworks: </span>\n "
+		for(var/datum/network/sub in N.sub)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+/datum/network_command/subnet/execute(datum/network/N, datum/network_argument/A, obj/item/device/hacktool/H)
+	..()
+	var/datum/network/SN
+	if(A.args[1] && N.sub.len >= 1)
```
```&& N.sub.len)``` instead of ```&& N.sub.len >= 1)```

nullbear replied:
a lot of it was kinda hastily coded at 1am, i've gone over it like 4 times so far cleaning and fixing stuff up. Thanks for pointing some of the stuff i missed out :)

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+	..()
+	var/datum/network/SN
+	if(A.args[1] && N.sub.len >= 1)
+		for(var/datum/network/S in N.sub)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+				return feed(H)
+	if(N.sub.len > 1 && !A.args[1])
+		var/not_stealthed = 0
+		for(var/datum/network/S in N.sub)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+				connect(S, H)
+				return feed(H)
+	if(N.sub.len > 1 && !A.args[1])
+		var/not_stealthed = 0
```
confusing var name, try ```avail_networks```

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+			connect(SN, H)
+			return feed(H)
+	if(N.sub.len == 1)
+		for(var/datum/network/S in N.sub)
```
I... what? N.sub only has one element, so why not just do ```SN = N.sub[1]```?

nullbear replied:
I... thanks.  I kinda felt dumb coding it but i couldnt think of how i should be doing it otherwise.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+		feedback = "<span class='warning'>Encryption Key Mismatch.</span>"
+		N.root.execute("security -alert -[N.id]", H) // Trigger the root's security. Quietly.
+		return feed(H)
+	for(var/datum/network/S in N.sub)
```
this should be an istypeless forloop.

nullbear replied:
what is an "istypeless forloop"

RemieRichards replied:
Check https://github.com/tgstation/-tg-station/blob/master/CONTRIBUTING.md#dream-maker-quirkstricks, the jist of it is that when you provide a typepath in a for each Byond will automatically ```istype()``` the list data for you, even if it's not necessary, such as a situation like this where ```N.sub``` will only contain network datums, so why istype them all?
It's a micro optimisation but it helps also to highlight that the list's data is all the same (or subtype'd)

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+	var/list/datum/network/found = list()
+	if(A.args[1] == "w" || A.args[1] == "wide")
+		if(!ishacktool(H))
+			for(var/datum/network/S in networks_by_wide)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,332 @@
+		feedback = "<span class='notice'>No Networks found within scope provided.</span>"
+		return feed(H)
+	feedback = "<span class='notice'>Probed Networks: </span>\n "
+	for(var/datum/network/S in found)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/hacktool.dm:
```diff
@@ -0,0 +1,37 @@
+	currfeed = F
+
+/obj/item/device/hacktool/proc/connect(var/datum/network/N)
+	if(istype(N, /datum/network))
```
```if(istype(N))```

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+
+/datum/network_argument
+	var/main_command
+	var/list/args = list()
```
please rename this to ```net_args``` I am increasingly worried a Human/DM will confuse this with the inbuild proc var ```args```

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+	..()
+	src.holder = holder
+	if(id)
+		while(networks_by_id[id])
```
instead of this nonsense, you can just make ```id``` ```var/static/id``` and ```id++``` before using the ID, this ensures a unique (although linearly numeric) ID.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+		active_network_ids += id
+		if(wireless)
+			networks_by_wide += src
+	var/list/commandlist = commands
```
this is a convoluted way of doing this.

```
for(var/C in commands)
    if(ispath(C))
        commands -= C
        commands += new C()
```

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+
+/datum/network/Destroy()
+	holder = null
+	for(var/datum/network/S in sub)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+	if(root)
+		root.del_sub(src)
+	root = list()
+	for(var/datum/network_command/NC in commands)
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+	root = null
+
+/datum/network/proc/del_sub(datum/network/S) // Called when the subdirectory is destroyed.
+	if(S in sub)
```
this operation isn't needed, if you try to remove something from a list that doesn't contain it DM just silently fails, no errors, no warnings, no issues.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+	var/datum/network_argument/A = new(command)
+
+	var/list/data = list()
+	for(var/datum/network_command/NC in commands) // allows for adding 'universal' commands, which can be attached to any network object.
```
this should be an istypeless forloop.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+			NC.execute(src, A, H) // Commands can be hard-coded to the network object, or made into a 'universal' command, which can be called by any network object.
+			qdel(A)
+			return 1
+	return
```
trailing return, remove it please.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+		return
+	var/datum/network_argument/A = new(command)
+
+	var/list/data = list()
```
unused.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,119 @@
+	var/info = ""		// A feedback message that explains the usage of this command. For the noobs.
+
+/datum/network_command/proc/execute(datum/network/N, datum/network_argument/A, obj/item/device/hacktool/H)
+	if(!N || !A)
```
this proc does nothing anyway, so there's no point making sure it has valid data... to do nothing with.

nullbear replied:
I put it here so that i didnt have to type it ever again. It's used in all the commands as ..()

RemieRichards commented:
I am tired and my eyes hurt, there may be more issues than I've already pointed out.

Supermichael777 commented:
whut is this shit?

nullbear commented:
@RemieRichards huge shoutout to you man, i wasn't really expecting anyone to go through this until i called for removal of the WIP or DNM tags.

@supermichael777 I'll update the description in a sec. I was in the middle of moving to a new house when i made the PR

nullbear commented:
@RemieRichards 
I can't find splittext after the removal of 509 support.

Iamgoofball commented:
@nullbear we replaced it with a new proc, check the 510 file

PKPenguin321 commented:
![?](http://www.pentadact.com/wp-content/Gunpoint-screenshots-01.png)

nullbear commented:
^ pretty much, though the networks will typically be much more convoluted than that.

Bawhoppen commented:
so like, logic gates and stuff going to be added?

Iamgoofball commented:
👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit👌👀👌👀👌👀👌👀👌👀 good shit go౦ԁ sHit👌 thats ✔ some good👌👌shit right👌👌there👌👌👌 right✔there ✔✔if i do ƽaү so my self 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯 👌👌 👌НO0ОଠOOOOOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ👌 👌👌 👌 💯 👌 👀 👀 👀 👌👌Good shit

Core0verload commented:
Goof is shit.

Cheridan commented:
Should probably call it something else, since i believe baystation already has a traitor item as a hacktool and it might get confusing

nullbear commented:
but this is /tg/station .-.

this hacktool probably does everything bays hacktool does, and more.

octareenroon91 commented:
How will this compare to goonstation's networking/radio-control mechanics?

nullbear commented:
I've never played on goon, and wouldnt know. But from what i could understand from what was discussed in coderbus irc, it's much more dynamic and intuitive. 

> Can you hack into, and bolt down an entire departments doors remotely? 

Yes.

If you have any suggestions for things you think should be possible using the network/hacktool, or that you want to be able to do, or just ideas on how you might use this in a typical round, please feel free to add a comment [here](https://docs.google.com/document/d/1wdT8FbhyK3t31SxKMVks6ielfZ-0nTLWlL0xtATyn9A/edit?usp=sharing).

RemieRichards commented:
@nullbear no worries, I love code review.

RemieRichards commented on code/modules/network/commands/directory_commands.dm:
```diff
@@ -0,0 +1,258 @@
+			return
+	feedback = "Invalid Argument: \"[A[1]]\""
+	feed(H)
+	return
```
trailing return, remove it please.
there's quite a few of these, I won't comment them all (unless requested)

nullbear replied:
Yeah i've honestly been using them for organisation, as it shows up as bright orange for me against a dark background, and makes it easier for me to look through my messy code. I'll remove them all eventually.

RemieRichards commented on code/modules/network/network.dm:
```diff
@@ -0,0 +1,190 @@
+		arguments[i-1] = butchered[i]
+	return list(command, arguments)
+
+proc/getuniqueid()
```
a lot more complicated that what I suggested really.

nullbear replied:
I haven't actually touched these up yet.

RemieRichards replied:
>splittext proc
See also: findtext proc jointext proc nonspantext proc spantext proc Regular expressions Format: splittext(Text,Delimiter,Start=1,End=0,include_delimiters=0) Returns: A list of text strings split by the delimiter given. Args: Text: The text string to search. Delimiter: A text string that will be used as the separator between items, OR a regular expression (regex) used to find splits Start: The text character position in Text in which to begin. End: The text character position in Text immediately following the last character to include in the splitting process. include_delimiters: True if any delimiters found should be included in the result. 

You need to be on 510, we're not 509 compat anymore.

nullbear replied:
i am on 510, else it wouldn't compile <_<

optimumtact replied:
https://tgstation13.org/ref/510/

optimumtact commented:
The networks are wireless at the moment I assume?

Is the intent to add a physical infrastructure for these in the form of repeaters or base stations or anything, or will it remain purely a simulated feature?

nullbear commented:
Yes they are wireless. I talked about having wired networks and did a poll on it, the typical response was "its future. everything is wireless" and complaints about how it would be a pain in the ass to do repairs and stuff. So i've decided to go with typically simulated/wireless networks.

There will probably be machines that add a physical networking aspect, as well as relays that allow for inter-z-level networking.

nullbear committed:
Adds network module

KorPhaeron commented:
A wired network would be fucking miserable for map making, already have to deal with wires/pipes/disposals everywhere

Core0verload commented:
Not if we use existing powernets.

nullbear commented:
^ the issue with that is that youd have to wire machines/cameras etc. I've done the polls, and it's been pretty much unanimously agreed that wired networks are not the way to go. It's also more resource intensive.

Networks have four remote modes, though most are typically only capable of 3. Proximity (need to be next to the machine in order to access the network.), Area (need to be in the same room to access the network. Like APC's), Stationwide (need to be on the same z level to access the network. Like telecomms) and Worldwide (you can access the network from anywhere. Intended to be used by cargo/thunderdome/telecomm relays etc.)

Linking machines can be done with a multitool, or with the hacktool. Typically it's best to use the hacktool for the increased options. Assuming the networks are both proximity and not remote, as it provides the most security, you'd have to physically connect to one network, and then go to the other machine in order to link them.

Networks are already gunna be slightly painful for mapping, hopefully i'll be able to implement some stuff to make it easier.

nullbear committed:
updates network module

GunHog commented:
Can you make this compile? I tried to test it and failed :(

Core0verload commented:
Wiring is OK as long as it uses APCs. APC connected - area is connected. This re-utilizes all the existing wiring and requires no map pain.

nullbear commented:
Sure gunhog just a sec, you'll need to update from master as well.

Also @coreoverload, how about apc's generating the networks based on wiring, and only receiving power through wires. But after wires are cut, the networks are maintained.

nullbear commented:
@gunhog there is currently nothing that uses networks, and no gui for the hacktool. So the only way to test it is to use advanced proc calls and stuff. ^^;

nullbear committed:
removes a bunch of global vars.

nullbear committed:
Adds key hacktool function.

It's "TECHNICALLY" functional now, but you'll have to use advanced proc
calls and VV in order to use it.

Just do hacktool/proc("command goes here")
and refresh vv for feedback messages.

Note, you'll still have to spawn a network.

nullbear committed:
a bit of work on camera networks

nullbear committed:
Adds a network var to machinery, since machinery will almost always have a network.

drynwyn commented:
If the hacktool doesn't look like a cyberdeck, you have done something wrong

Also this is the best thing, you're doing god's work null.

Razharas commented:
Real cool shit but you need to finish it @nullbear

nullbear commented:
it's gunna be a while. Atm im hitting a stub since i suck at html/tgui, and nobodies been responding to my hails for help on it.

Shadowlight213 commented:
@neersighted ^

nullbear commented:
like what needs to be done is pretty straightforward. I just dunno howwwww

nullbear commented:
not-abandoned so chill. Will re-open when it gets closer to a fully implemented system.

Jazaen commented:
THEY ALL SAY THAT

GunHog commented:
I really wanted this....

nullbear commented:
@GunHog I didnt really have much design solidified for it, so like 70% of what needs to be done currently is all design, and 30% is implementation.

GunHog commented:
Just...do not give up on it...this is too good to surrender....

Shadowlight213 commented:
RIP.
